### ISSUE TITLE
One apiserver can destabilize apiservers for all clusters in the given seed

### ISSUE THREAD

**Issue by dharapvj** [kubermatic/kubermatic#13321]
### What happened?

For one of the client, one cluster's apiserver pods got overloaded (presumably for network throughput but can also be for memory). The tale-tell signs were
1. Loads of context cancelled errors in rest of the control-plane components when trying to get/put queries to to apiserver like controller-manager, csi-driver etc.
2. Loads of context cancelled errors in kyverno admission controller running in cluster.
3. Many pods are in crashloop backoff and almost all of them are trying to do things with apiserver and apiserver not able to respond in time
4. konnectivity-server container in apiserver responds a LOT of logs like "Receive channel from agent is full" [Ref](https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/master/pkg/server/server.go#L784-L785)
5. Memory (via VPA) for apiserver had also shot up

Now Main issue here is that while all this was true for one cluster.. this impacted apiservers of ALL the other clusters managed via this seed! All the other clusters also observed timeout errors for apiserver but the network throughput increase was only on one cluster.

Check the graph:
![image](https://github.com/kubermatic/kubermatic/assets/1920926/11a9fd6a-4009-489c-9153-ca9ac662c30e)

### Expected behavior
One cluster's issues should not impact other clusters in anyway.

### How to reproduce the issue?
1. Create 3-4 of user clustrs
2. Deploy a purposefully nasty operator which will overwhelm apiserver of one cluster
3. Observe cluster operations in other cluster like new workload creation etc

### How is your environment configured?

- KKP version: 2.24.12
- Shared or separate master/seed clusters?: shared master/seed as well as dedicated seeds but issue seen here in dedicate seed.

### Provide your KKP manifest here (if applicable)

<details>

```yaml
# paste manifest here
```

</details>

### What cloud provider are you running on?

vsphere

### What operating system are you running in your user cluster?

Ubuntu 22.04

### Additional information

Additional information in this [internal slack thread](https://kubermatic.slack.com/archives/C01EWQZEW69/p1713511206777239)

**Comment by dharapvj** [comment]
queries to bring network utilization in above graph
```
A: sum by(namespace) (rate(container_network_receive_bytes_total{namespace=~"cluster-.*",pod=~"apiserver-.*"}[$__rate_interval]))
B: sum by (namespace) (rate(container_network_transmit_bytes_total{namespace=~"cluster-.*",pod=~"apiserver-.*"}[$__rate_interval]))
```

**Comment by cnvergence** [comment]
I have opened the PR related to this issue: https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/624
Issue: https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/586

**Comment by toschneck** [comment]
@cnvergence the PR seams to be merged, what is the next step to solve the issue?
cc @csengerszabo

**Comment by cnvergence** [comment]
@toschneck we need to wait for the release to make use of it, in the meantime, we could check out some other possible solutions, e.g rate-limiting has less impact from one user-cluster.

**Comment by toschneck** [comment]
@csengerszabo do we have this issue in mind for the next KKP release?
cc @dharapvj I think the issue is still present or?

**Comment by dharapvj** [comment]
Yes this is still potentially an issue. The logging in this area is quite less so we cannot say for certain that this is the cause of the issue with 100% reliability though.

**Comment by SimonTheLeg** [comment]
> With regards to the "receive channel from agent is full"

The change mentioned above will be available in Kubernetes 1.31., meaning clusters would need to be upgraded 1.31 to take effect. If I read the PR correctly it only makes the size configurable, so I assume you need to upgrade the cluster as well as set the `--xfr-channel-size` to a higher number (default is 10).

> Does it help to scale the apiserver replilcas from 2 to 3 or 4?

In theory it should. Reasoning behind this is that more apiserver replicas = more konnectivity servers increasing the total number of available channel sizes through an increased number of channels. I am purposely saying "in theory" because we have not good way to reproduce it at the moment. It's definitely a stop-gap solution and being able to set it directly in k8s 1.31 is much better, but it is worth a try.
If you decide to try it, I would kindly ask to share the results in here. This helps us tremendously in understanding the issue, especially since it is tough to reproduce atm.

**Comment by steled** [comment]
Thank you for the explanations.

What you think would be a good value for `--xfr-channel-size` in Kubernetes 1.31 to fix this problem?

**Comment by steled** [comment]
Just a short update from our side.
After we updated the apiserver replicas from 2 to 4 it looks, at least for the "Receive channel from agent is full" messages a bit better, see below (left side is before and right side after update):

![Image](https://github.com/user-attachments/assets/35273738-13f2-4f31-bb7d-d280bc9b5494)

**Comment by dharapvj** [comment]
@cnvergence - can you tell me - if user-clusters are at 1.31 (and KKP 2.26+), how to now override the channel-size?

**Comment by dharapvj** [comment]
yes exactly, I think we need to have a way in cluster spec to override this value? otherwise our changes will get overwritten by seed-controller-manager

**Comment by csengerszabo** [comment]
/label customer-request
/label sig/networking
Internal reference: 7777

**Comment by csengerszabo** [comment]
As agreed we'll move forward with this

> We can think to add some built-in overrides for Konnectivity in KKP, like specifying custom args to deployment/container.

CC: @cnvergence @buraksekili

**Comment by toschneck** [comment]
Finally some progress, please ensure that we also set the default in a way that covers the issue. Only the option itself, is likely won't be ware of

**Comment by SimonTheLeg** [comment]
quick question to @buraksekili @cnvergence 
This arg is only available for clusters with k8s version >= 1.31, right? 
Asking because documentation is still TBD, and if it is the case, we should definitely write down this information there

**Comment by buraksekili** [comment]
thanks for checking @SimonTheLeg! Yes, as far as I know, the Konnectivity fix (`--xfr-channel-size` flag) is included in Kubernetes v1.31.
I have updated https://github.com/kubermatic/kubermatic/pull/14189's documentation section. Please let me know if there's anything else I should do.

**Comment by cnvergence** [comment]
We can also create a doc section here https://docs.kubermatic.com/kubermatic/main/architecture/known-issues/, describing the issue and how to set these flags.

**Comment by SimonTheLeg** [comment]
@buraksekili ah just noticed, the `documentation` field in Pull requests does not work this way. It is intended to remind the PR author to write documentation and put a link to a docs PR (99% of the time a PR from here https://github.com/kubermatic/docs repository). So it works differently than the `release-notes` field, which actually gets parsed and written into the release notes automatically. All a bit unintuitive at first 😅 

What you need to do is create a pull request in the docs repo (I think @cnvergence mentioned a good place), and then put the link of your docs PR into the `documentation` field.

**Comment by vitality411** [comment]
> Finally some progress, please ensure that we also set the default in a way that covers the issue. Only the option itself, is likely won't be ware of

@buraksekili Have you decided against setting the default in a way that covers the issue? cc @toschneck 

Changing the default Konnectivity Server configuration is possible with the Seed’s defaultComponentSettings but changing the Konnectivity Agent configuration is only possible for each cluster individually. Would it not make sense to define default values for Server and Agent in the seed object and allow to overwrite these in the cluster object, as with other overrides?
https://github.com/kubermatic/docs/blob/5d2af0b922b2a4c1f0705763a04de210aaf0f4ff/content/kubermatic/main/data/seed.yaml#L447

**Comment by buraksekili** [comment]
Hi @vitality411, thank you for your input and for raising this!

We actually decided not to set a default value for the `--xfr-channel-size` argument because this configuration is only available to clusters running Kubernetes v1.31+, so introducing a default may cause issues during cluster upgrades/downgrades.


> changing the Konnectivity Agent configuration is only possible for each cluster

Actually, there is an alternative approach to apply default values across clusters: using a `ClusterTemplate`.

Here’s an example of a `ClusterTemplate` snippet that configures the `--xfr-channel-size` argument for the Konnectivity Agent:

```
apiVersion: kubermatic.k8c.io/v1
kind: ClusterTemplate
metadata:
spec:
  componentsOverride:
    konnectivityProxy:
      args:
      - --xfr-channel-size
      - "264"
```
When a cluster is created using this template, the Konnectivity Agent will be automatically configured with the specified value. You can confirm it with the following command:

```
$ kubectl --kubeconfig <cluster_cfg> get deployments -n kube-system konnectivity-agent -o yaml | grep -i "xfr" -A 1
        - --xfr-channel-size
        - "264"
```

However, if you think this is not needed/missing or if you have a use case where this doesn't solve your need, please feel free to open a GitHub issue and share more details with us. Thanks again for reaching out!

**Comment by vitality411** [comment]
@buraksekili We do not use cluster templates. I was hoping for a simple solution to increase the resilience of the environment by customizing the default server and agent configuration for existing and future clusters. I would appreciate if you could offer an alternative to the ClusterTemplates approch, otherwise I would have to solve this differently for future clusters.

**Comment by buraksekili** [comment]
Thank you @vitality411 for your request! I've raised a feature request #14782.

You're more than welcome to share any comments, suggestions, feedback, or use cases in that issue
